### ISSUE TITLE
symlink confusion

### ISSUE THREAD

**Issue by lucs** [rakudo/rakudo#5345]
⟨file-link⟩.IO.f reports True instead of False when ⟨file-link⟩ is a
symlink to a regular file.

Similarly ⟨dir-link⟩.IO.d reports True instead of False when
⟨dir-link⟩ is a symlink to a directory. If there was a predicate to
identify sockets (not sure why ⋯.IO.s was used to report the size of a
file, but there it is, tsk, tsk), the same problem would probably
occur.

Also, ⟨some-link⟩.IO.inode reports the inode number of the target
file, directory, or socket instead of the ⟨some-link⟩'s one.

Running in Linux with:

    Welcome to Rakudo™ v2023.02.
    Implementing the Raku® Programming Language v6.d.
    Built on MoarVM version 2023.02.

Example program illustrating the problems:

    #!/usr/bin/env raku

    my $targ = '/tmp/baz.targ';
    my $link = '/tmp/baz.link';

        # Create the $targ file and make $link a symlink to it.
    $targ.IO.spurt: 'baz';
    symlink $targ, $link;

        # Incorrectly reports .f as True, should be False. The same
        # problem occurs with a symlink to a directory when using
        # ⋯.IO.d.
    say "$link .l: ", $link.IO.l;
    say "$link .f: ", $link.IO.f;

        # Incorrectly reports $link's inode number as the same as
        # $targ's.
    say "$_ inode: {.IO.inode}" for $targ, $link;

        # The actual inode numbers.
    shell "ls -i $_" for $targ, $link;

        # Clean up.
    .IO.unlink for $targ, $link;

Prints this on my system:

    /tmp/baz.link .l: True
    /tmp/baz.link .f: True
    /tmp/baz.targ inode: 5243094
    /tmp/baz.link inode: 5243094
    5243094 /tmp/baz.targ
    5243095 /tmp/baz.link

**Comment by Leont** [comment]
This is correct behavior. That said, we could 

What we really need is a `stat` and an `lstat` so that people can be precise in what they want.

**Comment by lizmat** [comment]
I wonder if we could access the stat buffer completely using `nqp::dispatch`.  If that is possible, then I guess we need a `Stat` class with access methods matching the `STAT_` constants at https://github.com/Raku/nqp/blob/master/docs/ops.markdown#const

**Comment by Leont** [comment]
It's also useful for efficiency purposes, right now it takes two stat calls to check size an permissions, using a stat buffer you'd only need to do that once.

**Comment by lizmat** [comment]
Agree.  And also less chance of race conditions.

**Comment by niner** [comment]
We already have a stat buffer in MVMStat representation that is returned by the file-stat syscall. What's missing is a syscall to read the size from that stat buffer. We already have syscalls for several flags and times.

**Comment by Leont** [comment]
> We already have a stat buffer in MVMStat representation that is returned by the file-stat syscall. What's missing is a syscall to read the size from that stat buffer. We already have syscalls for several flags and times.

Yeah but it also needs a file-lstat (which shouldn't be hard to implement)
